docs: upfront recommend against using the no-return-await rule (#19727)
eslint/eslint: docs/src/rules/no-return-await.md

- https://jakearchibald.com/2017/await-vs-return-vs-return-await/
---

It is NOT recommended to use the `no-return-await` rule anymore because:

* `return await` on a promise will not result in an extra microtask.
* `return await` yields a better stack trace for debugging.

> Historical context: When promises were first introduced, calling `return await` introduced an additional microtask, one for the `await` and one for the return value of the async function. Each extra microtask delays the computation of a result and so this rule was added to help avoid this performance trap. Later, [V8 changed the way](https://v8.dev/blog/fast-async) `return await` worked so it would create a single microtask, which means this rule is no longer necessary.

Reviewer: "V8 changed the way" might imply that in other engines there can still be an extra microtask, depending on the implementation. This wasn't just an arbitrary decision or optimization in V8, but a change in the ECMAScript specification (https://github.com/tc39/ecma262/pull/1250), so I think we should keep "JavaScript now handles native Promises differently" or something like that.

> Historical context: When promises were first introduced, calling `return await` introduced an additional microtask, one for the `await` and one for the return value of the async function. Each extra microtask delays the computation of a result and so this rule was added to help avoid this performance trap. Later, [V8 changed the way](https://v8.dev/blog/fast-async) `return await` worked so it would create a single microtask, which means this rule is no longer necessary.

Reviewer 2: ```suggestion
Historical context: When promises were first introduced, calling `return await` introduced an additional microtask, one for the `await` and one for the return value of the async function. Each extra microtask delays the computation of a result and so this rule was added to help avoid this performance trap. Later, [ECMA-262 changed the way](https://github.com/tc39/ecma262/pull/1250) `return await` worked so it would create a single microtask, which means this rule is no longer necessary.
```

## Rule Details

> This rule warns on any usage of `return await`.

Reviewer: ```suggestion
This rule warns on any usage of `return await` except in `try` blocks.
```

- https://jakearchibald.com/2017/await-vs-return-vs-return-await/
---

It is NOT recommended to use the `no-return-await` rule anymore because:

* `return await` on a promise will not result in an extra microtask.

> * `return await` yields a better stack trace for debugging.

Reviewer 3: Much of this feels redundant with the paragraph 

> The original intent of this rule was to discourage the use of `return await`, to avoid an extra microtask. However, due to the fact that JavaScript now handles native `Promise`s differently, there is no longer an extra microtask. More technical information can be found in [this V8 blog entry](https://v8.dev/blog/fast-async).

Can you think of a way to combine them / remove the redundancy?

Reviewer 2: I agree. Let's just remove the paragraph. The bulleted list is all we need.

Author: Paragraph removed, [commented here](https://github.com/eslint/eslint/pull/19727#discussion_r2089604493).  Going to mark this thread as resolved.

> The following documentation is kept for historical context, and is not necessarily accurate anymore.

Reviewer 3: (praise) I think the callout that the rest of the docs page contains inaccurate information is a very good addition. It's annoyed me for a while that this page contained out of date information without such a disclaimer.

(suggestion) In order to be clear where the inaccurate information starts, what do you think about adding a horizontal line after this paragraph (a `***` in markdown)?

Reviewer 2: I think we can just be plain here.

```suggestion
Historical context: When promises were first introduced, calling `return await` introduced an additional microtask, one for the `await` and one for the return value of the async function. Each extra microtask delays the computation of a result and so this rule was added to help avoid this performance trap. Later, V8 changed the way `return await` worked so it would create a single microtask, which means this rule is no longer necessary.
```

Author: I like the paragraph that explains things for anyone curious. I've updated the PR to include the paragraph. Since the paragraph following this change was felt to be redundant (which I agree), I've re-included the link to the V8 blog entry into the suggested paragraph for anyone that wants to dive deeper.

> The following documentation is kept for historical context, and is not necessarily accurate anymore.

Reviewer 3: (optional) Important components about the docs are certainly not accurate, so you could go stronger on this, rather than hedging.

```suggestion
The following documentation is kept for historical context, and is not accurate anymore.
```

Author: I've went with the [suggested paragraph](https://github.com/eslint/eslint/pull/19727#discussion_r2089125889) which does not include that everything after is not accurate. If you feel there's value in having it, let me know.

> 

Reviewer 4: @kirkwaiblinger's done a ton of work on the tseslint side. I think it would be good to get Kirk's thoughts here if possible.

> 

Reviewer 4: @kirkwaiblinger's done a ton of work on the tseslint side. I think it would be good to get Kirk's thoughts here if possible.

Author: Thanks Josh for help pinging the right people. Appreciate it.

- https://jakearchibald.com/2017/await-vs-return-vs-return-await/
---

It is NOT recommended to use the `no-return-await` rule anymore because:

* `return await` on a promise will not result in an extra microtask.
* `return await` yields a better stack trace for debugging.

Historical context: When promises were first introduced, calling `return await` introduced an additional microtask, one for the `await` and one for the return value of the async function. Each extra microtask delays the computation of a result and so this rule was added to help avoid this performance trap. Later, [V8 changed the way](https://v8.dev/blog/fast-async) `return await` worked so it would create a single microtask, which means this rule is no longer necessary.

***

This rule warns on any usage of `return await`.

> Using `return await` inside an `async function` keeps the current function in the call stack until the Promise that is being awaited has resolved, at the cost of an extra microtask before resolving the outer Promise. `return await` can also be used in a try/catch statement to catch errors from another function that returns a Promise.

Reviewer 2: One last fix

```suggestion
Using `return await` inside an `async function` keeps the current function in the call stack until the Promise that is being awaited has resolved. `return await` can also be used in a try/catch statement to catch errors from another function that returns a Promise.
```

Author: Updated, removed mention of the extra microtask https://github.com/eslint/eslint/pull/19727/commits/d71bf2e7f725c62d2d29cda35555e79e99671d77

- https://jakearchibald.com/2017/await-vs-return-vs-return-await/
---

It is NOT recommended to use the `no-return-await` rule anymore because:

* `return await` on a promise will not result in an extra microtask.
* `return await` yields a better stack trace for debugging.

Historical context: When promises were first introduced, calling `return await` introduced an additional microtask, one for the `await` and one for the return value of the async function. Each extra microtask delays the computation of a result and so this rule was added to help avoid this performance trap. Later, [V8 changed the way](https://v8.dev/blog/fast-async) `return await` worked so it would create a single microtask, which means this rule is no longer necessary.

***

This rule warns on any usage of `return await`.

Using `return await` inside an `async function` keeps the current function in the call stack until the Promise that is being awaited has resolved. `return await` can also be used in a try/catch statement to catch errors from another function that returns a Promise.

> You can avoid the extra microtask by not awaiting the return value, with the trade off of the function no longer being a part of the stack trace if an error is thrown asynchronously from the Promise being returned. This can make debugging more difficult.

Author: Yeah, I do feel like maybe we just leave the inaccuracies.

- I think if you end up rewriting it to be accurate, you might end up with a very confusing doc.
- I think also because the rule is deprecated, and maybe will be removed in the future, that rewriting the doc is a poor use of time.
- I think to keep within the scope of the intention of this PR, of just warning people up front, what a rewrite consists of would warrant a much bigger discussion.

I think preserving the old, and calling it a day is what I'm leaning towards 👍.

If everyone is in agreement, my actions would be:

- Revert the previous commit https://github.com/eslint/eslint/commit/d71bf2e7f725c62d2d29cda35555e79e99671d77
- Add `The following documentation is kept for historical context, and is not accurate anymore.` before the horizontal line.

At that point, I think we're in a good place to wrap this up.

How does that sound @nzakas?

> You can avoid the extra microtask by not awaiting the return value, with the trade off of the function no longer being a part of the stack trace if an error is thrown asynchronously from the Promise being returned. This can make debugging more difficult.

Reviewer 2: @mikedidomizio I disagree. I think it's easy to miss that one line saying the following isn't accurate. If you'd prefer not to continue working on the PR, that's okay. I can take it over.

> You can avoid the extra microtask by not awaiting the return value, with the trade off of the function no longer being a part of the stack trace if an error is thrown asynchronously from the Promise being returned. This can make debugging more difficult.

Author: I'd hate to push work onto anyone, but you probably know what you want and how it'll be structured, to avoid the back and forth, it's probably best that you take over for the time being. 

I'm fine with you pushing commits onto this PR, or just sharing your proposed changes first for feedback. I'd love to see what you're thinking in how it'll look.
